File: app/javascript/channels/room_karma_channel.js
import consumer from "./consumer"

export default consumer.subscriptions.create("RoomKarmaChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const voteBtn = document.querySelector(`button[data-id="${data.id}"]`)
    if (voteBtn) {
      const spans = voteBtn.parentNode.getElementsByTagName("span")
      if (spans.length > 0) {
        spans[0].textContent = data.karma
      }
    }
  },

  upvoteMessage(data) {
    this.perform("upvote_message", {
      user_id: data.user,
      reviewer_id: data.reviewer,
      message_id: data.message,
    })
  },
})

File: app/channels/room_karma_channel.rb
class RoomKarmaChannel < ApplicationCable::Channel
  def subscribed
    current_user.rooms.each do |room|
      stream_from "rooms:#{room.id}:karma"
    end
  end

  def unsubscribed
    stop_all_streams
  end

  def upvote_message(data)
    message = Message.find(data['message_id'])
    return unless message

    calc_karma(message, data)
    VoteRelayJob.perform_later(message)
  end

  private

  def calc_karma(message, data)
    karma = Karma.find_by(user_id: data['user_id'],
                          reviewer_id: data['reviewer_id'],
                          message_id: message.id)

    if karma
      karma.destroy
    else
      save_karma(message, data)
    end
  end

  def save_karma(message, data)
    Karma.create(user_id: data['user_id'],
                 reviewer_id: data['reviewer_id'],
                 message_id: message.id,
                 point: 1)
  end
end
